io module stuff (#2357)
cpinitiative/usaco-guide: content/1_General/Input_Output.mdx

using namespace std;

int main() {
	int a;
	int b;
	int c;
	/*
	 * %d specifies that a value of type int is being input.
	 * To input a 64-bit (long long) number,
	 * use %lld instead (some OJs might need %I64 instead).
	 * Many other specifiers are also available; see link for more details.
	 * 
	 * Be sure to add a & character (address-of operator) when using
	 * scanf, UNLESS you are inputing a string with %s.
	 * 
	 * It is possible to input multiple values at a time as shown below.
	 */
	scanf("%d %d %d", &a, &b, &c);

	/*
	 * Specifiers for printf are mostly the same as those used
	 * by scanf, with the notable exception of floating-point numbers.
	 * Use a backslash character followed by the lowercase
	 * letter n to denote a newline.
	 * The address-of operator (&) is not used here.
	 */

> 	printf("The sum of these three numbers is %d\n", a+b+c);

Reviewer: ```suggestion
	printf("The sum of these three numbers is %d\n", a + b + c);
```
optional spacing stuff

Author: thx

```py
import sys

> # Read in integers n and m, both on the same line

Reviewer 2: ```suggestion
# Read in integers n and m, both on the same line, with list comprehension
```

Author: hold on actually
```suggestion
# Read in integers n and m on the same line with a list comprehension
```

n, m = [int(x) for x in input().split()]

> # This does the same thing

Reviewer 2: ```suggestion
# Read in integers n and m, but with list comprehension
```
these two are optional ^^

Author: ```suggestion
# Do the same thing but with map instead
```

```java
import java.io.*;

public class Main {
	static BufferedReader r = new BufferedReader(new InputStreamReader(System.in));

> 	static PrintWriter pw = new PrintWriter(System.out);

Reviewer 3: maybe move these into the main method as well (no reason for them to be static)

Author: yeah i thought that too, just didn't wanna change the code too much

Author: done

import java.util.*;

public class Main {
	public static void main(String[] args) {
		Kattio io = new Kattio();
		long x = io.nextLong();
		while (x != 1) {
			io.print(x + " ");
			if (x % 2 == 0) {
				x /= 2;
			} else {
				x = 3 * x + 1;
			}
		}
		io.println(x);
		// This line is needed, as it closes IO and flushes the output too
		io.close();
	}

> 	CodeSnip{Kattio}

Reviewer 3: ```suggestion
	// CodeSnip{Kattio}
```

	freopen("problemname.in", "r", stdin);
	// the following line creates/overwrites the output file
	freopen("problemname.out", "w", stdout);

> 	// cin now reads from the file instead of standard input

Reviewer 2: ```suggestion
	// cin now reads from the input file instead of standard input
```

	int a;
	int b;
	int c;
	cin >> a >> b >> c;

> 	// cout now prints to the file instead of standard output

Reviewer 2: ```suggestion
	// cout now prints to the output file instead of standard output
```

> 		int a = Integer.parseInt(st.nextToken());

Reviewer 3: Can you move the buffered reader and printwriter in here? it avoids doing all this try catch stuff up above

import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		Kattio io = new Kattio("problemname");
		int a = io.nextInt();
		int b = io.nextInt();
		int c = io.nextInt();
		io.print("The sum of these three numbers is ");
		io.println(a + b + c);
		// This line closes IO and flushes the output stream
		io.close();
	}

> 	CodeSnip{Kattio}

Reviewer 3: ```suggestion
	// CodeSnip{Kattio}
```

		int a = io.nextInt();
		int b = io.nextInt();
		int c = io.nextInt();

> 		io.print(""The sum of these three numbers is ");

Reviewer: ```suggestion
		io.print("The sum of these three numbers is ");
```

but use the built-in functions `.readline()` or `.readlines()`:
	
```py
"""
Note: The second argument can be omitted in the open()
command for read-only files
"""
fin = open("problemname.in", "r") 
fout = open("problemname.out", "w")

# One way to read the file using .readline()
line1 = fin.readline()
# readline() will pick up where you left off 
line2 = fin.readline()
line3 = fin.readline()

# Another way is to use a for loop and .readlines()
line_list = []
for line in fin.readlines():

> 	pass  # Do what you need to do here

Reviewer 2: ```suggestion
	pass  # Process input here
```
not sure if this is unclear (optional)

# printing line_list would give [line1, line2, line3]			

# Output:

> fout.write(output_text)  # Write to the output file

Reviewer 2: did you mean to put two spaces here? if so, add space for the other comment too

Author: thx ryan

Author: wait what i don't see it in my file

Reviewer 2: oh wait nvm, you deleted the other one. im blind

using namespace std;

int main() {
	// Use standard input to read from "paint.in"

> 	freopen("paint.in","r",stdin);

Reviewer: ```suggestion
	freopen("paint.in", "r", stdin);
```
optional spacing stuff

> 	// Use standard out to write to "paint.out"

Reviewer 4: standard output

> 	freopen("paint.out","w",stdout);

Reviewer: ```suggestion
	freopen("paint.out", "w", stdout);
```

int main() {
	ifstream fin("paint.in");
	ofstream fout("paint.out");

	vector<bool> cover(100);

> 	int a, b, c, d;

Reviewer 2: put this on separate lines to be consistent with java code and code before

			pw = new PrintWriter(new FileWriter("paint.out"));
		} catch(IOException e) {}
	}

	public static void main(String[] args) throws IOException {
		int a = r.nextInt();
		int b = r.nextInt();
		int c = r.nextInt();
		int d = r.nextInt();

		int[] cover = new int[100];
		for (int i = a; i < b; i++) {
			cover[i] = 1;
		}
		for (int i = c; i < d; i++) {
			cover[i] = 1;

> 		}

Reviewer 2: would it be better to use a bool array?

Author: i mean you're adding the numbers later on
doesn't matter that much tbh

Reviewer 2: alr

import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException{
		Kattio io = new Kattio("paint");
		int a = io.nextInt();
		int b = io.nextInt();
		int c = io.nextInt();
		int d = io.nextInt();

		int[] cover = new int[100];
		for (int i = a; i < b; i++) {
			cover[i] = 1;
		}
		for (int i = c; i < d; i++) {
			cover[i] = 1;
		}

		int ans = 0;
		for (int i = 0; i < cover.length; i++) {
			ans += cover[i];
		}
		io.println(ans);
		io.close();
	}

> 	CodeSnip{Kattio}

Reviewer 3: ```suggestion
	// CodeSnip{Kattio}
```

Author: wait what's the difference
does this break the build or smth

<Spoiler title="Method 1">

```py
with open('paint.in', 'r') as inp:
	lines = [line for line in inp]
	a, b = map(int, lines[0].split())
	c, d = map(int, lines[1].split())

	cover = [0] * 100
	for i in range(a, b):
		cover[i] = 1
	for i in range(c, d):
		cover[i] = 1
	
	ans = 0
	for i in range(len(cover)):
		ans += cover[i]
	with open('paint.out', 'w') as out:

> 		print(ans,file=out)  # out.write(str(ans) + '\n') also works

Reviewer: ```suggestion
		print(ans, file=out)  # out.write(str(ans) + '\n') also works
```
optional spacing :|

### Python

```py

> print(ans,end='')  # OK, no newline

Reviewer 4: ```suggestion
print(ans, end='')  # OK, no newline
```
for consistency
